### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by Zelinskyi-Serhii**
[DEMO LINK](https://Zelinskyi-Serhii.github.io/react_todo-app-add-and-delete/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,24 +1,258 @@
...
+import { Form } from './components/Form';
+import { ErrorMessage } from './components/ErrorMessage';
 
-const USER_ID = 0;
+const USER_ID = 6894;
 
-export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
+function filterTodosByCompleted(todos: Todo[], filterBy: string): Todo[] {
```
Comment (timoxa2t) [inline src/App.tsx:20]: make filterBy type FilterBy

File: src/App.tsx [hunk 2]
```diff
@@ -1,24 +1,258 @@
...
+              ));
+
+              setAllTodos(todosWithoutDeleted);
+            });
+        } catch {
+          setErrorMessage('Unable to delete a todo');
+        } finally {
+          setDeletedTodoIds([]);
+        }
+      }, 300);
```
Comment (timoxa2t) [inline src/App.tsx:88]: why do you need timer?

File: src/App.tsx [hunk 3]
```diff
@@ -1,24 +1,258 @@
...
```
Comment (timoxa2t) [inline src/App.tsx:117]:
```suggestion
     handleRemoveTodo(todo.id)
```

File: src/App.tsx [hunk 4]
```diff
@@ -1,24 +1,258 @@
...
+  }, [allTodos, filterType]);
+
+  const countActiveTodos = useMemo(() => {
+    return visibleTodos.reduce((sum, todo) => {
+      if (!todo.completed) {
+        return sum + 1;
+      }
+
+      return sum;
+    }, 0);
```
Comment (timoxa2t) [inline src/App.tsx:132]:
```suggestion
    return filterTodosByCompleted(allTodos, FilterBy.Active).length
```

File: src/App.tsx [hunk 5]
```diff
@@ -1,24 +1,258 @@
...
+  }, [visibleTodos]);
```
Comment (timoxa2t) [inline src/App.tsx:133]:
```suggestion
  }, [allTodos]);
```

File: src/App.tsx [hunk 6]
```diff
@@ -1,24 +1,258 @@
...
+
+  const isComplitedTodo = allTodos.some(todo => todo.completed);
+  const isActiveTodo = allTodos.some(todo => !todo.completed);
```
Comment (timoxa2t) [inline src/App.tsx:136]: I gues this is working fine but why do unnesessery array operations?
```suggestion
  const isComplitedTodo = allTodos.length - countActiveTodos > 0;
  const isActiveTodo = countActiveTodos > 0;
```

File: src/App.tsx [hunk 7]
```diff
@@ -1,24 +1,258 @@
...
+          removeTodo(id)
+            .then(() => {
+              const todosWithoutDeleted = allTodos.filter((todo) => (
+                todo.id !== id
+              ));
+
+              setAllTodos(todosWithoutDeleted);
+            });
+        } catch {
+          setErrorMessage('Unable to delete a todo');
```
Comment (timoxa2t) [inline src/App.tsx:84]:
```suggestion
          addError('Unable to delete a todo');
```

Comment (timoxa2t) [reply src/App.tsx:84]: I tried to simulate delete error and I didnt saw it. Can you check it and correct me if i'm wrong?

File: src/App.tsx [hunk 8]
```diff
@@ -1,24 +1,258 @@
...
+        return;
+      }
+
+      const newTodo = {
+        userId: USER_ID,
+        title: todoTitle,
+        completed: false,
+      };
+
+      setTempTodos({ ...newTodo, id: 0 });
```
Comment (timoxa2t) [inline src/App.tsx:153]:
```suggestion
      const newTodo = {
        userId: USER_ID,
        title: todoTitle,
        completed: false,
         id: 0,
      };

      setTempTodos(newTodo);
```

File: src/App.tsx [hunk 9]
```diff
@@ -1,24 +1,258 @@
...
+            </span>
+
+            <nav className="filter">
+              <TodoFilter
+                filteredTodos={filterType}
+                setFilteredTodos={setFilterType}
+              />
+            </nav>
+
+            {isComplitedTodo
```
Comment (timoxa2t) [inline src/App.tsx:237]: When you remove this button from footer, other footer child elements will shift to the right. You Should make it invisible instead. Look how its done in example

File: src/components/ErrorMessage/ErrorMessage.tsx [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  ({
+    errorMessage,
+    setErrorMessage,
+  }) => {
+    return (
+      <div className={
+        classNames(
+          'notification is-danger is-light has-text-weight-normal',
+          {
+            hidden: !errorMessage,
```
Comment (timoxa2t) [inline src/components/ErrorMessage/ErrorMessage.tsx:19]: hidden is not hiding element instantly. it fades away, and we cant remove error text while it fades away. But its just my opinion)

**Review by timoxa2t — CHANGES_REQUESTED** [review]
Sorry, but you have to make some changes. I gues it would be easier if you move code from App to come Additional components

File: src/App.tsx (hunk 1 above)
Comment (Zelinskyi-Serhii) [reply src/App.tsx:20]: I already have this type, i just forgot)

File: src/App.tsx (hunk 2 above)
Comment (Zelinskyi-Serhii) [reply src/App.tsx:88]: I have bugs and i created this crutche)

File: src/App.tsx [hunk 11]
```diff
@@ -1,24 +1,247 @@
...
+
+export const App: FC = () => {
+  const [allTodos, setAllTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [todoTitle, setTodoTitle] = useState('');
+  const [filterType, setFilterType] = useState<FilterBy>(FilterBy.ALL);
+  const [tempTodos, setTempTodos] = useState<Todo | null>(null);
+  const [activeInput, setActiveInput] = useState(true);
+  const [isQuery, setIsQuery] = useState(false);
+  const [deletedTodoIds, setDeletedTodoIds] = useState<number[]>([]);
```
Comment (OleksiiBlezniuk) [inline src/App.tsx:47]: Isn't this state used for showing loader? In that case rename it to `loadingTodoIds`

File: src/App.tsx [hunk 12]
```diff
@@ -1,24 +1,247 @@
...
+            const todosWithoutDeleted = allTodos.filter(({ completed }) => (
+              !completed
+            ));
+
+            setAllTodos(todosWithoutDeleted);
+          });
+      } catch {
+        setErrorMessage('Unable to delete a todo');
+      } finally {
+        setDeletedTodoIds([]);
```
Comment (OleksiiBlezniuk) [inline src/App.tsx:111]: this will reset all loading todos when at least one of them will be removed

Comment (OleksiiBlezniuk) [reply src/App.tsx:111]: if you want to disable all loading todos at one time 

1.  pass all ids to array before running async function for each todo
2. replace `forEach` with map
3. use Promise.all for this map
4. await this promise
5. then set empty array to loading todo ids

File: src/App.tsx [hunk 13]
```diff
@@ -1,24 +1,247 @@
...
+      addTodo(newTodo)
+        .then(todo => {
+          setAllTodos(todos => [...todos, todo]);
+        })
+        .catch(() => addError('Unable to add a todo'))
+        .finally(() => {
+          setActiveInput(true);
+          setTodoTitle('');
+          setTempTodos(null);
+        });
```
Comment (OleksiiBlezniuk) [inline src/App.tsx:155]: Let's use async/await

File: src/App.tsx [hunk 14]
```diff
@@ -1,24 +1,247 @@
...
+}
+
+export const App: FC = () => {
+  const [allTodos, setAllTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [todoTitle, setTodoTitle] = useState('');
+  const [filterType, setFilterType] = useState<FilterBy>(FilterBy.ALL);
+  const [tempTodos, setTempTodos] = useState<Todo | null>(null);
+  const [activeInput, setActiveInput] = useState(true);
+  const [isQuery, setIsQuery] = useState(false);
```
Comment (OleksiiBlezniuk) [inline src/App.tsx:46]: what is isQuery?

File: src/App.tsx [hunk 15]
```diff
@@ -1,24 +1,247 @@
...
+  todo: Todo,
+  isLoading: boolean,
+}
+
+export const App: FC = () => {
+  const [allTodos, setAllTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [todoTitle, setTodoTitle] = useState('');
+  const [filterType, setFilterType] = useState<FilterBy>(FilterBy.ALL);
+  const [tempTodos, setTempTodos] = useState<Todo | null>(null);
```
Comment (OleksiiBlezniuk) [inline src/App.tsx:44]: but why it's called todo**s** if it's only one?

File: src/App.tsx (hunk 9 above)
Comment (OleksiiBlezniuk) [reply src/App.tsx:226]: Agree

**Review by illiarerun — APPROVED** [review]

**Pull request closed without merging**